src/Stratis.Bitcoin/Base/ChainRepository.cs

```diff
@@ -17,23 +18,108 @@ public interface IChainRepository : IDisposable
         Task SaveAsync(ConcurrentChain chain);
     }
 
-    public class ChainRepository : IChainRepository
+    public interface IFinalizedBlockHeight : IDisposable
     {
+        /// <summary>Gets the finalized block height.</summary>
+        /// <returns>Height of a block that can't be reorged away from.</returns>
+        int GetFinalizedBlockHeight();
+
+        /// <summary>Saves the finalized block height.</summary>
+        /// <param name="height">Block height.</param>
+        Task SaveFinalizedBlockHeightAsync(int height);
+    }
+
+    public class ChainRepository : IChainRepository, IFinalizedBlockHeight
+    {
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
         /// <summary>Access to DBreeze database.</summary>
         private readonly DBreezeEngine dbreeze;
 
         private BlockLocator locator;
 
-        public ChainRepository(string folder)
+        /// <summary>Database key under which the block hash of the coin view's current tip is stored.</summary>
+        private static readonly byte[] finalizedBlockKey = new byte[0];
+
+        /// <summary>Height of a block that can't be reorged away from.</summary>
+        private int finalizedHeight;
+
+        public ChainRepository(string folder, ILoggerFactory loggerFactory)
         {
             Guard.NotEmpty(folder, nameof(folder));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
 
             this.dbreeze = new DBreezeEngine(folder);
+
+            this.finalizedHeight = this.LoadFinalizedBlockHeightAsync().GetAwaiter().GetResult();
```

Comment:
This should probably happen in the `Load()` we need to try keep constructors as light as possible as they are used by the DI.

Reply:
Do you think it should be done on the Chain load or as a separated Load that is called from the base feature prior to Chain loading?

Reply:
Added as a separated value.